tracing-opentelemetry/src/layer.rs

```diff
@@ -163,37 +202,131 @@ where
     /// let tracer = provider.get_tracer("component-name");
     ///
     /// // Create a layer with the configured tracer
-    /// let otel_layer = OpenTelemetryLayer::with_tracer(tracer);
+    /// let otel_layer = tracing_opentelemetry::layer().with_tracer(tracer);
+    ///
+    /// // Use the tracing subscriber `Registry`, or any other subscriber
+    /// // that impls `LookupSpan`
+    /// let subscriber = Registry::default()
+    ///     .with(otel_layer);
+    /// # drop(subscriber);
+    /// ```
+    pub fn with_tracer<Tracer>(self, tracer: Tracer) -> OpenTelemetryLayer<S, Tracer>
+    where
+        Tracer: api::Tracer + 'static,
+    {
+        OpenTelemetryLayer {
+            tracer,
+            sampler: self.sampler,
+            id_generator: self.id_generator,
+            get_context: WithContext(OpenTelemetryLayer::<S, Tracer>::get_context),
+            _registry: self._registry,
+        }
+    }
+
+    /// Set the [`Sampler`] to configure the logic around which [`Span`]s are
+    /// exported.
+    ///
+    /// [`Sampler`]: https://docs.rs/opentelemetry/latest/opentelemetry/api/trace/sampler/trait.Sampler.html
+    /// [`Span`]: https://docs.rs/opentelemetry/latest/opentelemetry/api/trace/span/trait.Span.html
+    ///
+    /// # Examples
+    ///
+    /// ```rust,no_run
+    /// use opentelemetry::sdk;
+    /// use tracing_subscriber::layer::SubscriberExt;
+    /// use tracing_subscriber::Registry;
+    ///
+    /// // The probability sampler can be used to export a percentage of spans
+    /// let sampler = sdk::Sampler::Probability(0.33);
+    ///
+    /// // Create a layer with the configured sampler
+    /// let otel_layer = tracing_opentelemetry::layer().with_sampler(sampler);
     ///
     /// // Use the tracing subscriber `Registry`, or any other subscriber
     /// // that impls `LookupSpan`
     /// let subscriber = Registry::default()
     ///     .with(otel_layer);
     /// # drop(subscriber);
     /// ```
-    pub fn with_tracer(tracer: T) -> Self {
+    pub fn with_sampler<Sampler>(self, sampler: Sampler) -> Self
+    where
+        Sampler: api::Sampler + 'static,
+    {
+        OpenTelemetryLayer {
+            sampler: Box::new(sampler),
+            ..self
+        }
+    }
+
+    /// Construct a new layer with the specified [`Tracer`] and [`Sampler`].
+    ///
+    /// [`Tracer`]: https://docs.rs/opentelemetry/latest/opentelemetry/api/trace/tracer/trait.Tracer.html
+    /// [`Sampler`]: https://docs.rs/opentelemetry/latest/opentelemetry/api/trace/sampler/trait.Sampler.html
+    fn new<Sampler>(tracer: T, sampler: Sampler) -> Self
```

Comment:
Do we want this constructor to be public as well?

Reply:
Might as well be, sure.